File: backend/app/api/api_v1/endpoints/kanban.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from uuid import uuid4

from app.db.base import get_db
from app.schemas.kanban import (
    KanbanBoard, KanbanBoardCreate, KanbanBoardUpdate,
    KanbanColumn, KanbanColumnCreate, KanbanColumnUpdate
)
from app.models.kanban import KanbanBoard as KanbanBoardModel, KanbanColumn as KanbanColumnModel
from app.core.security import get_current_user
from app.models.user import User

router = APIRouter()

DEFAULT_COLUMNS = [
    {"title": "Inbox", "color": "#4A90E2", "is_static": True},
    {"title": "To Do", "color": "#555555", "is_static": False},
    {"title": "In Progress", "color": "#888888", "is_static": False},
    {"title": "Done", "color": "#cccccc", "is_static": False},
]


# ==================== BOARDS ====================

@router.get("/boards", response_model=List[KanbanBoard])
def read_boards(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get boards for user"""
    boards = db.query(KanbanBoardModel).filter(
        KanbanBoardModel.user_id == current_user.id
    ).order_by(KanbanBoardModel.board_order).all()
    
    return boards


@router.post("/boards", response_model=KanbanBoard, status_code=201)
def create_board(
    *,
    db: Session = Depends(get_db),
    board_in: KanbanBoardCreate,
    current_user: User = Depends(get_current_user),
):
    """Create new board"""
    # Get max board_order
    max_order = db.query(KanbanBoardModel.board_order).filter(
        KanbanBoardModel.user_id == current_user.id
    ).order_by(KanbanBoardModel.board_order.desc()).first()
    
    order = (max_order[0] + 1) if max_order and max_order[0] is not None else 0
    
    board = KanbanBoardModel(
        id=str(uuid4()),
        title=board_in.title,
        board_order=order,
        user_id=current_user.id
    )
    db.add(board)
    db.commit()
    db.refresh(board)
    return board


@router.put("/boards/{board_id}", response_model=KanbanBoard)
def update_board(
    *,
    db: Session = Depends(get_db),
    board_id: str,
    board_in: KanbanBoardUpdate,
    current_user: User = Depends(get_current_user),
):
    """Update board"""
    board = db.query(KanbanBoardModel).filter(
        KanbanBoardModel.id == board_id,
        KanbanBoardModel.user_id == current_user.id
    ).first()
    
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    
    update_data = board_in.model_dump(exclude_unset=True)
    # Rename 'order' to 'board_order' if present
    if 'order' in update_data:
        update_data['board_order'] = update_data.pop('order')
    for field, value in update_data.items():
        setattr(board, field, value)
    
    db.commit()
    db.refresh(board)
    return board


@router.delete("/boards/{board_id}")
def delete_board(
    *,
    db: Session = Depends(get_db),
    board_id: str,
    current_user: User = Depends(get_current_user),
):
    """Delete board"""
    board = db.query(KanbanBoardModel).filter(
        KanbanBoardModel.id == board_id,
        KanbanBoardModel.user_id == current_user.id
    ).first()
    
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    
    db.delete(board)
    db.commit()
    return {"message": "Board deleted successfully"}


@router.post("/boards/reorder")
def reorder_boards(
    board_ids: List[str],
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Reorder boards"""
    for index, board_id in enumerate(board_ids):
        board = db.query(KanbanBoardModel).filter(
            KanbanBoardModel.id == board_id,
            KanbanBoardModel.user_id == current_user.id
        ).first()
        if (board):
            board.board_order = index
    
    db.commit()
    return {"message": "Boards reordered successfully"}


# ==================== COLUMNS ====================

@router.get("/boards/{board_id}/columns", response_model=List[KanbanColumn])
def read_columns(
    board_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get columns for board"""
    # Check that board exists and belongs to user
    board = db.query(KanbanBoardModel).filter(
        KanbanBoardModel.id == board_id,
        KanbanBoardModel.user_id == current_user.id
    ).first()
    
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    
    columns = db.query(KanbanColumnModel).filter(
        KanbanColumnModel.board_id == board_id
    ).order_by(KanbanColumnModel.order).all()
    
    # Return columns without creating default ones
    # Boards are created empty now, users create their own columns
    return columns


@router.get("/inbox/categories", response_model=List[KanbanColumn])
def read_inbox_categories(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get Inbox category columns"""
    # Get all user's boards first
    user_board_ids = db.query(KanbanBoardModel.id).filter(
        KanbanBoardModel.user_id == current_user.id
    ).all()
    user_board_ids = [board.id for board in user_board_ids]
    
    # Get inbox categories (board_id is NULL) OR categories from user's boards
    columns = db.query(KanbanColumnModel).filter(
        (KanbanColumnModel.board_id == None) |
        (KanbanColumnModel.board_id.in_(user_board_ids))
    ).filter(
        KanbanColumnModel.is_inbox_category == True
    ).order_by(KanbanColumnModel.order).all()
    
    return columns


@router.post("/boards/{board_id}/columns", response_model=KanbanColumn, status_code=201)
def create_column(
    *,
    board_id: str,
    db: Session = Depends(get_db),
    column_in: KanbanColumnCreate,
    current_user: User = Depends(get_current_user),
):
    """Create new column"""
    # Check that board exists and belongs to user
    board = db.query(KanbanBoardModel).filter(
        KanbanBoardModel.id == board_id,
        KanbanBoardModel.user_id == current_user.id
    ).first()
    
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    
    # Get max order
    max_order = db.query(KanbanColumnModel.order).filter(
        KanbanColumnModel.board_id == board_id
    ).order_by(KanbanColumnModel.order.desc()).first()
    
    order = (max_order[0] + 1) if max_order and max_order[0] is not None else 0
    
    column = KanbanColumnModel(
        id=str(uuid4()),
        title=column_in.title,
        color=column_in.color,
        order=order,
        is_static=column_in.is_static,
        is_inbox_category=column_in.is_inbox_category,
        board_id=board_id
    )
    db.add(column)
    db.commit()
    db.refresh(column)
    return column


@router.post("/inbox/categories", response_model=KanbanColumn, status_code=201)
def create_inbox_category(
    *,
    db: Session = Depends(get_db),
    column_in: KanbanColumnCreate,
    current_user: User = Depends(get_current_user),
):
    """Create new Inbox category column"""
    # Get max order for Inbox categories (board_id is NULL)
    max_order = db.query(KanbanColumnModel.order).filter(
        KanbanColumnModel.board_id == None,
        KanbanColumnModel.user_id == current_user.id
    ).order_by(KanbanColumnModel.order.desc()).first()
    
    order = (max_order[0] + 1) if max_order and max_order[0] is not None else 0
    
    column = KanbanColumnModel(
        id=str(uuid4()),
        title=column_in.title,
        color=column_in.color,
        order=order,
        is_static=column_in.is_static,
        is_inbox_category=True,
        board_id=None,  # NULL для Inbox категорий
        user_id=current_user.id  # Устанавливаем user_id для Inbox категорий
    )
    db.add(column)
    db.commit()
    db.refresh(column)
    return column


@router.put("/columns/{column_id}", response_model=KanbanColumn)
def update_column(
    *,
    db: Session = Depends(get_db),
    column_id: str,
    column_in: KanbanColumnUpdate,
    current_user: User = Depends(get_current_user),
):
    """Update column"""
    column = db.query(KanbanColumnModel).join(KanbanBoardModel).filter(
        KanbanColumnModel.id == column_id,
        KanbanBoardModel.user_id == current_user.id
    ).first()
    
    if not column:
        raise HTTPException(status_code=404, detail="Column not found")
    
    update_data = column_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(column, field, value)
    
    db.commit()
    db.refresh(column)
    return column


@router.delete("/columns/{column_id}")
def delete_column(
    *,
    db: Session = Depends(get_db),
    column_id: str,
    current_user: User = Depends(get_current_user),
):
    """Delete column"""
    column = db.query(KanbanColumnModel).join(KanbanBoardModel).filter(
        KanbanColumnModel.id == column_id,
        KanbanBoardModel.user_id == current_user.id
    ).first()
    
    if not column:
        raise HTTPException(status_code=404, detail="Column not found")
    
    if column.is_static:
        raise HTTPException(status_code=400, detail="Cannot delete static column")
    
    db.delete(column)
    db.commit()
    return {"message": "Column deleted successfully"}


@router.post("/columns/reorder")
def reorder_columns(
    column_ids: List[str],
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Reorder columns"""
    for index, column_id in enumerate(column_ids):
        column = db.query(KanbanColumnModel).join(KanbanBoardModel).filter(
            KanbanColumnModel.id == column_id,
            KanbanBoardModel.user_id == current_user.id
        ).first()
        if (column):
            column.order = index
    
    db.commit()
    return {"message": "Columns reordered successfully"}

File: frontend/src/stores/kanban.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import axios from 'axios'

interface KanbanBoard {
  id: string
  title: string
  order: number
  user_id: string
}

interface KanbanColumn {
  id: string
  title: string
  order: number
  color: string
  is_static: boolean
  is_inbox_category: boolean
  board_id: string | null
}

export type InboxCategory = KanbanColumn

export const useKanbanStore = defineStore('kanban', () => {
  const boards = ref<KanbanBoard[]>([])
  const columns = ref<KanbanColumn[]>([])
  const currentBoardId = ref<string | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  const api = axios.create({
    baseURL: '/api/v1/kanban',
    headers: {
      'Content-Type': 'application/json'
    }
  })

  // ==================== BOARDS ====================

  const fetchBoards = async () => {
    loading.value = true
    error.value = null
    try {
      const response = await api.get('/boards')
      boards.value = response.data
      
      // Set current board to first board
      if (!currentBoardId.value) {
        currentBoardId.value = boards.value[0]?.id || null
      }
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка загрузки досок'
      console.error('Error fetching boards:', err)
    } finally {
      loading.value = false
    }
  }

  const createBoard = async (boardData: Partial<KanbanBoard>) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.post('/boards', boardData)
      boards.value.push(response.data)
      return response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка создания доски'
      console.error('Error creating board:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateBoard = async (boardId: string, boardData: Partial<KanbanBoard>) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.put(`/boards/${boardId}`, boardData)
      const index = boards.value.findIndex(b => b.id === boardId)
      if (index !== -1) {
        boards.value[index] = response.data
      }
      return response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка обновления доски'
      console.error('Error updating board:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  const deleteBoard = async (boardId: string) => {
    loading.value = true
    error.value = null
    try {
      await api.delete(`/boards/${boardId}`)
      boards.value = boards.value.filter(b => b.id !== boardId)
      
      // Switch to another board if current board was deleted
      if (currentBoardId.value === boardId) {
        currentBoardId.value = boards.value[0]?.id || null
      }
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка удаления доски'
      console.error('Error deleting board:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  const reorderBoards = async (boardIds: string[]) => {
    try {
      await api.post('/boards/reorder', boardIds)
      boardIds.forEach((id, index) => {
        const board = boards.value.find(b => b.id === id)
        if (board) board.order = index
      })
      boards.value.sort((a, b) => a.order - b.order)
    } catch (err: any) {
      console.error('Error reordering boards:', err)
    }
  }

  const setCurrentBoard = async (boardId: string) => {
    currentBoardId.value = boardId
    columns.value = []
    if (boardId === '' || boardId === null) {
      await fetchInboxCategories()
    } else {
      await fetchColumns()
    }
  }

  const currentBoard = computed(() => {
    return boards.value.find(b => b.id === currentBoardId.value) || null
  })

  // ==================== COLUMNS ====================

  const fetchColumns = async () => {
    if (!currentBoardId.value) return
    
    loading.value = true
    error.value = null
    try {
      const response = await api.get(`/boards/${currentBoardId.value}/columns`)
      columns.value = response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка загрузки колонок'
      console.error('Error fetching columns:', err)
    } finally {
      loading.value = false
    }
  }

  const createColumn = async (columnData: Partial<KanbanColumn>) => {
    if (!currentBoardId.value) throw new Error('No board selected')
    
    loading.value = true
    error.value = null
    try {
      const response = await api.post(`/boards/${currentBoardId.value}/columns`, {
        ...columnData,
        board_id: currentBoardId.value
      })
      columns.value.push(response.data)
      return response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка создания колонки'
      console.error('Error creating column:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateColumn = async (columnId: string, columnData: Partial<KanbanColumn>) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.put(`/columns/${columnId}`, columnData)
      const index = columns.value.findIndex(c => c.id === columnId)
      if (index !== -1) {
        columns.value[index] = response.data
      }
      return response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка обновления колонки'
      console.error('Error updating column:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  const deleteColumn = async (columnId: string) => {
    loading.value = true
    error.value = null
    try {
      await api.delete(`/columns/${columnId}`)
      columns.value = columns.value.filter(c => c.id !== columnId)
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка удаления колонки'
      console.error('Error deleting column:', err)
      throw err
    } finally {
      loading.value = false
    }
  }

  const reorderColumns = async (columnIds: string[]) => {
    try {
      await api.post('/columns/reorder', columnIds)
      columnIds.forEach((id, index) => {
        const col = columns.value.find(c => c.id === id)
        if (col) col.order = index
      })
      columns.value.sort((a, b) => a.order - b.order)
    } catch (err: any) {
      console.error('Error reordering columns:', err)
    }
  }

  // ==================== INBOX CATEGORIES ====================
  
  const fetchInboxCategories = async () => {
    loading.value = true
    error.value = null
    try {
      const response = await api.get('/inbox/categories')
      columns.value = response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка загрузки Inbox категорий'
      console.error('Error fetching inbox categories:', err)
    } finally {
      loading.value = false
    }
  }
  
  const createInboxCategory = async (columnData: Partial<KanbanColumn>) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.post('/inbox/categories', columnData)
      columns.value.push(response.data)
      return response.data
    } catch (err: any) {
      error.value = err.response?.data?.detail || 'Ошибка создания Inbox категории'
      console.error('Error creating inbox category:', err)
      throw err
    } finally {
      loading.value = false
    }
  }
  
  return {
    boards,
    columns,
    currentBoardId,
    currentBoard,
    loading,
    error,
    fetchBoards,
    createBoard,
    updateBoard,
    deleteBoard,
    reorderBoards,
    setCurrentBoard,
    fetchColumns,
    createColumn,
    updateColumn,
    deleteColumn,
    reorderColumns,
    fetchInboxCategories,
    createInboxCategory
  }
})
